pages/DSNP/Graph.md

```diff
@@ -108,20 +108,38 @@ In the following section, the Alice to Bob identifier for context C is called PR
-1. Alice derives a context-specific subkey <code>k<sub>Bob</sub></code> from the shared secret `k` as the master key, using the Blake2b hash function as in the libsodium function `crypto_kdf_derive_from_key`, using Bob's DSNP User Id as the 64-bit key identifier and the ASCII encoding of the [PRId Context](#contexts) string (`"PRIdCtx0"` for connections) as the 8-byte human-readable identifier.
+Definitions:
+* <code>Id<sub>A</sub> = _DSNP User Id of A_</code>
+* <code>Id<sub>B</sub> = _DSNP User Id of B_</code>
+
+Algorithm:
+
+1. Both Alice and Bob generate an asymmetric key pair for use with x25519 <abbr title="Elliptic Curve Integrated Encryption Scheme">ECIES</abbr>.
+   Each publishes a Public Key Announcement with their generated public key with a `keyType` value of `keyAgreement`.
+    * Libsodium: [`crypto_box_keypair`](https://libsodium.gitbook.io/doc/public-key_cryptography/authenticated_encryption#key-pair-generation)
+    * <code>(A<sub>public</sub>, A<sub>private</sub>) = <abbr title="Key Generation Function">KGF</abbr>()</code>
+    * <code>(B<sub>public</sub>, B<sub>private</sub>) = <abbr title="Key Generation Function">KGF</abbr>()</code>
+1. When Alice wants to interact with Bob, she looks up Bob's public key and performs an x25519 Elliptic-curve Diffie-Hellman key exchange operation using her private key and Bob's public key, generating a root shared secret.
+    * Libsodium: [`crypto_box_beforenm`](https://libsodium.gitbook.io/doc/public-key_cryptography/authenticated_encryption#precalculation-interface)
+    * <code>RootSharedSecret<sub>AB</sub> = <abbr title="Elliptic-curve Diffie-Hellman">ECDH</abbr>(B<sub>public</sub>, A<sub>private</sub>)</code>
+1. Alice derives a context-specific subkey <code>CtxSharedSecret<sub>Bob</sub></code> from the shared secret `RootSharedSecret` as the master key, Bob's DSNP User Id as the 64-bit key identifier, and the ASCII encoding of the [PRId Context](#contexts) string (`"PRIdCtx0"` for connections).
+    * Libsodium: [`crypto_kdf_derive_from_key`](https://libsodium.gitbook.io/doc/key_derivation)
+    * <code>Ctx<sub>KDF</sub> = "PRIdCtx0"</code>
+    * <code>CtxSharedSecret<sub>A→B</sub> = <abbr title="Key Derivation Function">KDF</abbr>(Id<sub>B</sub> Ctx<sub>KDF</sub>, RootSharedSecret<sub>AB</sub>)</code>
 1. Alice uses Bob's DSNP User Id to form a 8-byte message.
-1. Alice encrypts the message using the PRId key <code>k<sub>Bob</sub></code> and a nonce of her own User Id (padded with zero bytes), as in the libsodium function `crypto_secretbox_easy`.
-1. Alice removes the message authentication code (MAC), leaving the 8-byte value <code>PRId<sub>A→B,C</sub></code>.
-  (Because Alice will publish the PRId, it can be authenticated without a MAC.)
+    * <code>MSG<    sub>A→B</sub> = Id<sub>B</sub></code>
+1. Alice encrypts the message using the PRId key <code>CtxSharedSecret<sub>A→B</sub></code> and a nonce of her own User Id (padded with zero bytes).
```

Comment:
what is the byte size of this padded nonce value? it would be nice to have an example of padding

Reply:
It is 24 bytes for XSalsa20 adding that in...

Reply:
Updated. No example, but a specific endianness as well so much better defined.